design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,286 @@
+
+To achieve this, we need to build a framework for exporting metrics from the Kanister controller, and to 
+start with, export some metrics to Prometheus. 
+
+Some example metrics include:
+ActionSets succeeded
+ActionSets failed
+Action duration
+Phase duration, etc.
+
+
+## Scope
+1. Design a framework that allows us to export new
+Kanister metrics to Prometheus easily. 
+2. Add a few fundamental metrics related to ActionSets and Blueprints to start with.
+
+
+## Detailed Design
+
+### Architecture
+
+![Alt text](Metrics_design.png?raw=true "Prometheus Integration Design")
+
+#### Text description
+1. The initializer of the consumer package calls new_metrics, a helper method that talks to Kanister’s metrics package. The result is a new metrics struct that owns all the prometheus collectors.
+
+2. In order to initialize all the required prometheus collectors, the new_metrics method calls the InitCounterVec, InitGaugeVec, InitHistogramVec in the metrics package. It passes the metric names and the specific label names and label values, aka, BoundedLabels, to the metrics package. Once it initializes all the prometheus collectors successfully, it returns an struct that wraps all the collectors, which the consumer package can then use. 
+
+3. The metrics package internally attempts to initialize the collectors and register the specific collectors with prometheus. If the registration fails because the specific metric with label header already exists, the collector will simply be returned to the caller. If the registration fails due to other reasons, then the metrics package will cause a panic, signalling programmer error. In case of the CounterVec, the InitCounterVec attempts to generate all possible permutations of label values and sets each counter 
+within the CounterVec to 0. 
+
+4. Once the collector is created in the metrics package, it will be returned to the consumer package’s new_metrics helper method.
+
+5. Once the initialization of all collectors are complete, a new metrics struct will be returned to the consumer’s package initializer. 
+
+6. Suppose the consumer package wants to increment a specific counter in a counter vec, it constructs a prometheus.Labels mapping using a helper method to retrieve the specific counter from the counter vec and performs an increment operation. 
+
+
+
+### Low Level APIs
+
+#### Metrics Package
+
+
+
+```golang
+// BoundedLabel is a type that represents a label and its associated 
+// valid values
+type BoundedLabel struct {
+	LabelName   string
+	LabelValues []string
+}
+```
+An example of a BoundedLabel is in the scenario of ActionSet resolutions.
+Suppose we want to track these resolutions across different blueprints, 
+we would create the bounded labels in the following way:
+```golang
+BoundedLabel{
+  LabelName: "operation_type"
+  LabelValues: ["backup", "restore"]
+}
+
+BoundedLabel{
+  LabelName: "action_set_resolution"
+  LabelValues: ["success", "failure"]
+}
```

Comment:
@miquella This isn't intended to be go code. It's just an example of how an instance of the struct will look like. Do we still need to maintain the indentation? Because the markdown isn't giving us the intention automatically. 
Please let me know if there are any better ways to represent this example.

Reply:
@mellon-collie: This issue is more that the rendered version of this document is inconsistent. This makes it more difficult to read the document and for no particular reason.

Furthermore, it may cause the reader to wonder why things are different and whether those differences important? (obviously indentation isn't important in Go, but distracting the reader limits their ability to consume and understand the document)